pkg/util/util_test.go

```diff
@@ -607,3 +607,82 @@ func TestIsIngressConfigInUse(t *testing.T) {
 		}
 	}
 }
+
+func TestConfigurationBackup(t *testing.T) {
+	testCases := []struct {
+		testName string
+		broker   v1beta1.Broker
+	}{
+		{
+			testName: "empty broker",
+			broker:   v1beta1.Broker{},
+		},
+		{
+			testName: "detailed broker",
+			broker: v1beta1.Broker{
+				Id:                0,
+				BrokerConfigGroup: "default",
+				ReadOnlyConfig: `advertised.listeners=INTERNAL://kafka-0.kafka.svc.cluster.local:9092
+broker.id=0
+cruise.control.metrics.reporter.bootstrap.servers=kafka-all-broker.kafka.svc.cluster.local:9092
+cruise.control.metrics.reporter.kubernetes.mode=true`,
+				BrokerConfig: &v1beta1.BrokerConfig{
+					Image:                "Image",
+					MetricsReporterImage: "MetricsReporterImage",
+					Config: `advertised.listeners=INTERNAL://kafka-0.kafka.svc.cluster.local:9092
+broker.id=0
+cruise.control.metrics.reporter.bootstrap.servers=kafka-all-broker.kafka.svc.cluster.local:9092
+cruise.control.metrics.reporter.kubernetes.mode=true`,
+					BrokerLabels: map[string]string{"apple": "tree"},
+					Affinity: &corev1.Affinity{
+						NodeAffinity: &corev1.NodeAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+								NodeSelectorTerms: []corev1.NodeSelectorTerm{
+									{
+										MatchExpressions: nil,
+										MatchFields: []corev1.NodeSelectorRequirement{
+											{
+												Key:      "apple",
+												Operator: "in",
+												Values:   []string{"fruit"},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					PodSecurityContext:   &corev1.PodSecurityContext{},
+					SecurityContext:      &corev1.SecurityContext{},
+					BrokerIngressMapping: []string{"apple"},
+					InitContainers: []corev1.Container{
+						{
+							Name:  "test-initcontainer",
+							Image: "busybox:latest",
+						},
+						{
+							Name:  "a-test-initcontainer",
+							Image: "test/image:latest",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range testCases {
+		config, err := GzipAndBase64BrokerConfiguration(&test.broker)
+		if err != nil {
+			t.Errorf("error should be nil, got: %v", err)
+		}
+
+		broker, err := GetBrokerFromBrokerConfigurationBackup(config)
+		if err != nil {
+			t.Errorf("error should be nil, got: %v", err)
+		}
+
+		if !reflect.DeepEqual(test.broker, broker) {
+			t.Errorf("Expected: %v  Got: %v", test.broker, broker)
+		}
```

Comment:
**Note**: we might want to separate these two unit test cases in the future and test the functions for their functionality, because right now we are only ensuring that the `GetBrokerFromBrokerConfigurationBackup()` function is the inverse of the `GzipAndBase64BrokerConfiguration()` function, but not that they do what we expect of them (gzip and base64 encoding the JSON broker configuration and doing the reverse for decoding).
For example you pass a pointer to the `test.broker` to the `GzipAndBase64BrokerConfiguration()` function which could in theory by accident overwrite the values of it to make it an empty `v1beta1.Broker{}` and `GetBrokerFromBrokerConfigurationBackup()` could just be returning empty `v1beta1.Broker{}` objects all the time and the `reflect.DeepEqual()` call would still be true, but the functions would not be doing what they are expected to do.

I do admit the functions seem to work correctly based on their current code, but the test itself does not guarantee to uphold that state in case the functions are modified in the future for any reason.

Reply:
Thank you for the explanation. This would be an edge case but you are right. I missed this case.